### PR TITLE
style: enable all biomejs recommended rules

### PR DESCRIPTION
**Pull request by nazarhussain**
**Motivation**

Make sure all recommended rules from biomejs are enabled. 

**Description**

- Enable all recommended rules 
- If some recommended rule is disabled, explain with proper reason

### CODE REVIEW

**Comment by nazarhussain** [comment]
Would be much easier to review commit by commit.

File: biome.jsonc
Comment (wemeetagain) [inline biome.jsonc]: comment is wrong

File: biome.jsonc
Comment (wemeetagain) [inline biome.jsonc]:
```suggestion
        // We use to import types and object without differentiating
```

File: biome.jsonc
Comment (wemeetagain) [inline biome.jsonc]: worth discussion on this one. Afaik, typescript can infer the type most of the time so its not really necessary.

File: biome.jsonc
Comment (wemeetagain) [inline biome.jsonc]: supposedly, the rule checks if the empty export is necessary or not, see the valid case https://biomejs.dev/linter/rules/no-useless-empty-export/

File: biome.jsonc
Comment (wemeetagain) [inline biome.jsonc]: we use named functions to ease with debugging, but that should be covered by the rule,  "Named function expressions are ignored" https://biomejs.dev/linter/rules/use-arrow-function/#_top

File: biome.jsonc
Comment (wemeetagain) [inline biome.jsonc]: this probably shouldn't be disabled?

File: biome.jsonc
Comment (wemeetagain) [inline biome.jsonc]: this rule should probably be enabled

File: biome.jsonc
Comment (wemeetagain) [inline biome.jsonc:67]: worth discussion to see if we want useExportType and useImportType

File: packages/beacon-node/src/eth1/eth1DataCache.ts
Comment (wemeetagain) [inline packages/beacon-node/src/eth1/eth1DataCache.ts]:
```suggestion
    return highestEth1Data?.blockNumber ?? null;
```

And a few other cases in this PR

**Review by wemeetagain — COMMENTED** [review]
generally looks good

File: packages/state-transition/src/util/blindedBlock.ts
Comment (nflaig) [inline packages/state-transition/src/util/blindedBlock.ts]: Not a fan of the `performance/noDelete` rule, setting a property to `undefined` is not equivalent in some cases

e.g.
```node
>  Object.keys({})
[]
>  Object.keys({foo: undefined})
[ 'foo' ]
```
or when merging objects
```node
> {...{hello: "world"}, ...{}}
{ hello: 'world' }
> {...{hello: "world"}, ...{hello: undefined}}
{ hello: undefined }
```

sometimes you just wanna get rid of the property altogether

File: packages/api/test/utils/utils.ts
Comment (nflaig) [inline packages/api/test/utils/utils.ts]: could just refactor this to `err != null`

File: packages/beacon-node/test/e2e/chain/bls/multithread.test.ts
Comment (nflaig) [inline packages/beacon-node/test/e2e/chain/bls/multithread.test.ts:43]: bit strange to move this into `beforeEach`

File: packages/light-client/src/spec/utils.ts
Comment (nflaig) [inline packages/light-client/src/spec/utils.ts:99]: I doubt this makes a difference where we put the `default` because it will have to evaluate all `case` either way

so just semantically it makes more sense to me to put the `default` last as it will always be called last if not `case` matches

File: packages/beacon-node/src/sync/sync.ts
Comment (nflaig) [inline packages/beacon-node/src/sync/sync.ts:167]: could also throw an error which includes the value, just in case this actually happens makes debugging easier

File: packages/state-transition/src/util/blindedBlock.ts
Comment (matthewkeil) [reply packages/state-transition/src/util/blindedBlock.ts]: I'm actually a fan of this one.  Not having delete makes turbofan optimization much better.  We should be paying attention to the usage.  I've noticed it when we go from one SSZ container to another and curious if there is a way to update the ssz to ignore `undefined` fields if they are not part of the container definition...

Comment (nflaig) [reply packages/state-transition/src/util/blindedBlock.ts]: > ssz to ignore undefined fields if they are not part of the container definitio

that's already the case, it only reads fields that are defined in the container, and ignores all others

> Not having delete makes turbofan optimization much better

need to carefully review all cases then that are in this PR, afterwards it should be fine, just don't like lint rules that change behavior

File: packages/beacon-node/src/sync/sync.ts
Comment (nazarhussain) [reply packages/beacon-node/src/sync/sync.ts:167]: The funny thing is Typescript complains `rangeSyncState.status` to be `never` type so can't be used.

File: biome.jsonc
Comment (nazarhussain) [reply biome.jsonc:67]: Created an issue to discuss this: https://github.com/ChainSafe/lodestar/issues/7159

File: packages/state-transition/src/util/blindedBlock.ts
Comment (nazarhussain) [reply packages/state-transition/src/util/blindedBlock.ts]: Reverted the change for this rule from this PR. 

I am in favor of not using `delete`, but this should be done with more deliberate review in a separate PR.

File: packages/light-client/src/spec/utils.ts
Comment (nazarhussain) [reply packages/light-client/src/spec/utils.ts:99]: From the usage it seems author want to fail early without evaluation of the rest of the code. As all of the cases have fall-through behavior and if any case match the `default` will execute as well.

File: biome.jsonc
Comment (nazarhussain) [reply biome.jsonc]: Done

File: biome.jsonc
Comment (nazarhussain) [reply biome.jsonc]: Done

File: biome.jsonc
Comment (nazarhussain) [reply biome.jsonc]: Done

File: packages/beacon-node/src/sync/sync.ts
Comment (nflaig) [reply packages/beacon-node/src/sync/sync.ts:167]: makes sense from typescript pov this can never happen, let's keep it as is

File: packages/light-client/src/spec/utils.ts
Comment (nflaig) [reply packages/light-client/src/spec/utils.ts:99]: [according to chatgpt](https://chatgpt.com/share/670cef74-1f48-8013-8360-18fc717f7c2f) the order of  `default` does not matter

> the default case will only be executed if no other case matches

File: packages/state-transition/src/util/blindedBlock.ts
Comment (nflaig) [reply packages/state-transition/src/util/blindedBlock.ts]: Let's do it in another PR if this has performance improvements we can take advantage of

File: biome.jsonc
Comment (nazarhussain) [reply biome.jsonc]: Done

File: biome.jsonc
Comment (nazarhussain) [reply biome.jsonc]: Done

File: packages/light-client/src/spec/utils.ts
Comment (nazarhussain) [reply packages/light-client/src/spec/utils.ts:99]: ChatGPT is wrong here. Check this [snippet](https://www.typescriptlang.org/play/?#code/MYewdgzgLgBAhgLjAVwLYCMCmAnGBeGAJgG4YAoMiAdwEspgALGACjgEoYBvMmXmYOBEwwAjAnJ9J-cBBAAbTADo5IAObMARCI1tiPPgKFFx+qdMjylK9RsI69kgCaYAZnGRyoJs71AWFymqazm4eUPamUmQAvkA).  and notice the logs.

Comment (nflaig) [reply packages/light-client/src/spec/utils.ts:99]: yeah, in case of fall throug the order matters, kinda makes sense

**Review by wemeetagain — COMMENTED** [review]
LGTM, waiting for another approval to merge

File: packages/beacon-node/src/chain/opPools/aggregatedAttestationPool.ts
Comment (nflaig) [inline packages/beacon-node/src/chain/opPools/aggregatedAttestationPool.ts:639]: do we really think this is a good change? Especially for larger code block being explicit with the `else` seems good to me as it does not rely on `if`-block to return or throw

File: packages/fork-choice/test/unit/protoArray/getCommonAncestor.test.ts
Comment (nflaig) [inline packages/fork-choice/test/unit/protoArray/getCommonAncestor.test.ts]: this looks wrong, if anything this should be

```suggestion
      if (ancestor !== null) {
```

File: packages/cli/src/applyPreset.ts
Comment (nflaig) [inline packages/cli/src/applyPreset.ts:95]: according to the comment these exports seem to have a purpose

Comment (nazarhussain) [reply packages/cli/src/applyPreset.ts:95]: If there is an import, we don't need an export to make it a module.

File: packages/beacon-node/src/chain/opPools/aggregatedAttestationPool.ts
Comment (nazarhussain) [reply packages/beacon-node/src/chain/opPools/aggregatedAttestationPool.ts:639]: Was suggested by @wemeetagain snd I am also in favor of not using such else bocks. More levels of indentations make code less readable.

File: packages/fork-choice/test/unit/protoArray/getCommonAncestor.test.ts
Comment (nazarhussain) [reply packages/fork-choice/test/unit/protoArray/getCommonAncestor.test.ts]: That's fine. Just avoiding matching undefined to the null.

Comment (nflaig) [reply packages/fork-choice/test/unit/protoArray/getCommonAncestor.test.ts]: what is if `ancestorNode` is null but it shouldn't be, we wanna match it against `ancestor`, this modifies the tests in a way were we lose confidence that `ancestorNode.blockRoot === ancestor`

File: packages/beacon-node/src/chain/opPools/aggregatedAttestationPool.ts
Comment (nflaig) [reply packages/beacon-node/src/chain/opPools/aggregatedAttestationPool.ts:639]: > More levels of indentations make code less readable.

I don't think this can be applied generally, some cases I think identation can help make it more readable

Comment (wemeetagain) [reply packages/beacon-node/src/chain/opPools/aggregatedAttestationPool.ts:639]: no strong opinions from me. I don't think that this needs to be enforced by the linter.

Comment (nflaig) [reply packages/beacon-node/src/chain/opPools/aggregatedAttestationPool.ts:639]: I would just go with the style that the original author of the code preferred, `else` is more explicit and can support readability imo.

Doesn't apply everything e.g. sometimes you just wanna throw an error and not use else, if it's a 1-liner

Comment (nazarhussain) [reply packages/beacon-node/src/chain/opPools/aggregatedAttestationPool.ts:639]: If indentation is needed for readability, one can use code blocks. There is a reason community named such style `noUselessElse`. 

Also if there is larger code block which needs indentation for readability, it should be splited into multiple functions.

Comment (matthewkeil) [reply packages/beacon-node/src/chain/opPools/aggregatedAttestationPool.ts:639]: I tend to agree with that... else is a code smell of overly long code blocks that could be made more readable...

**Review by nflaig — DISMISSED** [review]

File: packages/state-transition/src/util/blindedBlock.ts
Comment (matthewkeil) [reply packages/state-transition/src/util/blindedBlock.ts]: this may have unintended consequences... i forget the exact situation but the ssz library was not happy about creating a root (or maybe it was container statics accepting the value, but i forget) when there was an undefined after converting container types...

**Review by matthewkeil — APPROVED** [review]
LGTM 🎸

**Review by wemeetagain — APPROVED** [review]

**Review by nflaig — APPROVED** [review]

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.23.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.23.0) :tada:

**Pull request merged**
